src/zjs_fs.c

```diff
@@ -0,0 +1,940 @@
+
+        jerry_value_t error = jerry_create_number(ret);
+        zjs_signal_callback(id, &error, sizeof(jerry_value_t));
+    }
+
+    return ZJS_UNDEFINED;
+}
+
+static jerry_value_t zjs_unlink_sync(const jerry_value_t function_obj,
+                                    const jerry_value_t this,
+                                    const jerry_value_t argv[],
+                                    const jerry_length_t argc) {
+    return zjs_unlink(function_obj, this, argv, argc, 0);
+}
+
+static jerry_value_t zjs_unlink_async(const jerry_value_t function_obj,
+                                     const jerry_value_t this,
+                                     const jerry_value_t argv[],
+                                     const jerry_length_t argc) {
+    return zjs_unlink(function_obj, this, argv, argc, 1);
+}
+
+static jerry_value_t zjs_read(const jerry_value_t function_obj,
+                             const jerry_value_t this,
+                             const jerry_value_t argv[],
+                             const jerry_length_t argc,
+                             uint8_t async)
+{
+    file_handle* handle;
+    int err = 0;
+
+    if (argc < 6 && async) {
+        ERR_PRINT("not enough parameters\n");
+        return ZJS_UNDEFINED;
+    }
+
+    if (!jerry_value_is_object(argv[0])) {
+        ERR_PRINT("first parameter was not file descriptor\n");
+        return ZJS_UNDEFINED;
+    }
+    if (!jerry_value_is_object(argv[1])) {
+        ERR_PRINT("second parameter was not buffer\n");
+        return ZJS_UNDEFINED;
+    }
+    if (!jerry_value_is_number(argv[2])) {
+        ERR_PRINT("third parameter was not offset\n");
+        return ZJS_UNDEFINED;
+    }
+    if (!jerry_value_is_number(argv[3])) {
+        ERR_PRINT("fourth parameter was not length\n");
+        return ZJS_UNDEFINED;
+    }
+    if (!jerry_value_is_number(argv[4])) {
+        ERR_PRINT("fifth parameter was not position\n");
+        return ZJS_UNDEFINED;
+    }
+    if (async) {
+        if (!jerry_value_is_function(argv[5])) {
+            ERR_PRINT("sixth parameter was not callback function\n");
+            return ZJS_UNDEFINED;
+        }
+    }
+    if (!jerry_get_object_native_handle(argv[0], (uintptr_t*)&handle)) {
+        ERR_PRINT("native handle not found\n");
+        return ZJS_UNDEFINED;
+    }
+    zjs_buffer_t* buffer = zjs_buffer_find(argv[1]);
+    uint32_t offset = jerry_get_number_value(argv[2]);
+    uint32_t length = jerry_get_number_value(argv[3]);
+    uint32_t position = jerry_get_number_value(argv[4]);
+
+    // if a position was specified, seek to it before writing
+
+    if (fs_seek(&handle->fp, position, SEEK_SET) != 0) {
+        ERR_PRINT("error seeking to position\n");
+        return ZJS_UNDEFINED;
+    }
+
+    DBG_PRINT("reading into fp=%p, buffer=%p, offset=%lu, length=%lu\n", &handle->fp, buffer->buffer, offset, length);
+
+    uint32_t ret = fs_read(&handle->fp, buffer->buffer + offset, length);
```

Comment:
When you use JS Buffer APIs they check for overflow but right you have to do the checks yourself when using the buffer directly. As I mention below, maybe a helper function in buffer class could make it easier.